controllers/hcpvaultsecretsapp_controller.go

```
@@ -288,6 +295,66 @@ func (r *HCPVaultSecretsAppReconciler) Reconcile(ctx context.Context, req ctrl.R
+	ticker := time.NewTicker(1 * time.Hour)
+
+	// this process is expected to run indefinitely
+	for range ticker.C {
+		err := r.cleanupOrphanedShadowSecrets(ctx)
+		if err != nil {
+			logger.Error(err, "Failed to cleanup orphaned shadow secrets")
+		}
+	}
+}
+
+func (r *HCPVaultSecretsAppReconciler) cleanupOrphanedShadowSecrets(ctx context.Context) error {
+	logger := log.FromContext(ctx)
+
+	// filtering only for dynamic secrets
+	dynamicSecretLabelSelector := client.MatchingLabels{"app.kubernetes.io/component": "hvs-dynamic-secret-cache"}
+
+	// get all secrets in the operator namespace
+	secrets := corev1.SecretList{}
+	if err := r.List(ctx, &secrets, client.InNamespace(common.OperatorNamespace), dynamicSecretLabelSelector); err != nil {
+		logger.Error(err, "Failed to list secrets")
+	}
+
+	for _, secret := range secrets.Items {
+		o := &secretsv1beta1.HCPVaultSecretsApp{}
+
+		namespace := secret.Labels[hvsaLabelPrefix+"/namespace"]
+		name := secret.Labels[hvsaLabelPrefix+"/name"]
+		namespacedName := types.NamespacedName{Namespace: namespace, Name: name}
+
+		// get the HCPVaultSecretsApp instance that that the shadow secret belongs to (if applicable)
+		err := r.Get(ctx, namespacedName, o)
+		if err != nil && !apierrors.IsNotFound(err) {
+			logger.Error(err, "Error getting resource from k8s", "secret", secret.Name)
+			continue
+		}
+
+		if o.GetDeletionTimestamp() != nil && o.GetUID() == types.UID(secret.Labels[labelOwnerRefUID]) {
+			if err := r.handleDeletion(ctx, o); err != nil {
+				logger.Error(err, "Failed to handle deletion of HCPVaultSecretsApp", "app", o.Name)
+				return err
+			}
+
+			logger.Info("Deleted orphaned resources associated with HCPVaultSecretsApp", "app", o.Name)
+		} else if apierrors.IsNotFound(err) || secret.GetDeletionTimestamp() != nil {
+			if err := r.Client.Delete(ctx, &secret); err != nil {
+				logger.Error(err, "Failed to delete secret", "secret", secret.Name)
+				return err
```

It looks like an error on one deletion will prevent the pruning of other secrets.

I believe we already have some code for pruning the secret cache. I wonder if we could use that instead? 

https://github.com/hashicorp/vault-secrets-operator/blob/336f6ca2973424c1df5fec2b87701c06d8fcc193/vault/cache_storage.go#L144

---

Good catch! I think changing it to `continue` instead of returning would be the best approach because it will continue pruning secrets after the failed deletion of one, but it will eventually come back to the failed deletion during the next cleanup round.

---

`ClientCacheStorage` isn't being used at all in `hcpvaultsecretsapp_controller.go`, so I thought it would be easier to do it this way. If I were to add `ClientCacheStorage`, would it look like the following?

* Add `ClientCacheStorage` to `HCPVaultSecretsAppReconciler`
```
type HCPVaultSecretsAppReconciler struct {
	// ...
	ClientCacheStorage          vault.ClientCacheStorage
	StorageConfig               *vault.ClientCacheStorageConfig
}
```

* Initialize `ClientCacheStorage` in `SetupWithManager()`
```
func (r *HCPVaultSecretsAppReconciler) SetupWithManager(mgr ctrl.Manager, opts controller.Options) error {
// ...
storageConfig := vault.DefaultClientCacheStorageConfig()
	var metricsRegistry prometheus.Registerer
	if r.CollectClientCacheMetrics {
		metricsRegistry = ctrlmetrics.Registry
	}
	clientCacheStorage, err := vault.NewDefaultClientCacheStorage(context.Background(), r.Client, storageConfig, metricsRegistry)
	if err != nil {
		return fmt.Errorf("failed to create client cache storage: %w", err)
	}
	r.ClientCacheStorage = clientCacheStorage
}
```

And then simply use `Prune` from `ClientCacheStorage` in the cleanup method that I wrote, right?

---

Discussed with @benashz today and we decided that this change would be too involved for a simple ticket.